File: leetcodes/49. Group Anagrams/groupAnagrams.java
import java.util.Collection;
import java.util.LinkedList;

class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
    //iterate over strs
     //iterate over each string
        //put each letter into new char[26] array
     //set char[26] array as new string
      //char type 0~127 is enough for constraint 0 <= strs[i].length <= 100
       //6 ms use char(0~127) array and new String(frequencyArr) method.
     //create tempList as map.getOrDefault(key, new LinkedList<String>())
     //push s into tempList
     //put templist as value of key in map
    //retrun new List->map.values() -> value Set of map
    //time O(mxn) spaceO(k) -> worse case O(n)
    if (strs == null || strs.length == 0) new ArrayList<>();
    Map<String, List<String>> map = new HashMap<>();
    for (String s: strs) {
      char[] frequencyArr = new char[26];
      for (int i = 0; i < s.length(); i++) {
        frequencyArr[s.charAt(i) - 'a'] ++;
      }
      String key = new String(frequencyArr);
      List<String> tempList = map.getOrDefault(key, new LinkedList<String>());
      tempList.add(s);
      map.put(key, tempList);
    }
    return new LinkedList<>(map.values());
  }
}

File: leetcodes/49. Group Anagrams/groupAnagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        '''
        iterate over the string array
          compare each s in strs to see if they're anagrams
          a.sort -> need to create new str spaceO(n)and take O(nlogn) times
            easy to compare each string
          b. use char[26] to store the freq of each index(char) and convert it to string for comparision
            time o(n*m) spaceO(n)
          put it into map as key-> easy to search key for comparision, add the str into the values as an array
        return the values of map
        '''
        map = dict()
#        map.setdefault('', [])
        #map = collections.defaultdict(list) it can set default value as a list
        for s in strs:
          charFreq = [0] * 26
          for c in s:
            index = ord(c) - ord('a')
            charFreq[index] += 1
          key = tuple(charFreq) # tuple is immutable, and I can't use join to convert it to string since it is integer in array element
          if map.get(key):
            map[key].append(s)
          else:
            map[key] = [s]

        return map.values()
